[ao] Implemented InputWeightEqualization Detector observer insertion

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80191
* __->__ #79962

Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This implements the functionality of adding observer points for the
input-weight equalization detector and contains the relavent tests for
this functionality. The full Detector functionality will be fleshed out
in a later commit.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao][work-in-progress] Implemented InputWeightEqualization Detector

Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This code is still work in progress and has so far implemented finding
detector insert points, and is still needing generating report
funcitonality.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao][work-in-progress] Implemented InputWeightEqualization Detector"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This code is still work in progress and has so far implemented finding
detector insert points, and is still needing generating report
funcitonality.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao][work-in-progress] Implemented InputWeightEqualization Detector"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This code is still work in progress and has so far implemented finding
detector insert points, and is still needing generating report
funcitonality.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao][work-in-progress] Implemented InputWeightEqualization Detector"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This code is still work in progress and has so far implemented finding
detector insert points, and is still needing generating report
funcitonality.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented InputWeightEqualization Detector observer insertion"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This implements the functionality of adding observer points for the
input-weight equalization detector and contains the relavent tests for
this functionality. The full Detector functionality will be fleshed out
in a later commit.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented InputWeightEqualization Detector observer insertion"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This implements the functionality of adding observer points for the
input-weight equalization detector and contains the relavent tests for
this functionality. The full Detector functionality will be fleshed out
in a later commit.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented InputWeightEqualization Detector observer insertion"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This implements the functionality of adding observer points for the
input-weight equalization detector and contains the relavent tests for
this functionality. The full Detector functionality will be fleshed out
in a later commit.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -42,6 +41,34 @@ def get_detector_name(self) -> str:
+            raise ValueError("passed in target_fqn not found in graph's targets.")
+        else:
+            # recursively call it with parent fqn
+            return self._get_targeting_node(prepared_fx_model, target_fqn[:parent_fqn_sep_index])
```
i'm not sure how this is supposed to work.

if we have a simple example without fusion, it looks fine, but i'm not sure how the additional code solves the fusion case.

so lets say we have 

model:
-submodel0 = random_op - linear - relu
-submodel1 = conv - relu

and we are searching for fqn submodel0.linear's node, but that was fused with submodel0.relu, when we hit the node we actually want its node.target is now submodel0.linear.0 so when we compare that (in line 60) with target_fqn, we don't see equality, so then we go to line 70 and recurse using fqn=submodel0 which i'm not sure how that helps us both because in this example, we're not getting the input to linear, we're getting the input to submodel0/random_op and because if we can't find the linear, we won't be able to find its weight.

can you maybe explain whats going on here better, the comments are a bit sparse.

vspenubarthi replied:
So I just ran the test case `TestFxDetectInputWeightEqualization.test_input_weight_equalization_report_gen` with print statements and here is what's going on:
The module right as it enters the prepare_callibration on line 1141 of test_model_report_fx:
```
GraphModule(
  (activation_post_process_0): HistogramObserver()
  (block1): Module(
    (linear): LinearReLU(
      (0): Linear(in_features=3, out_features=3, bias=True)
      (1): ReLU()
    )
    (conv2d): ConvReLU2d(
      (0): Conv2d(3, 3, kernel_size=(1, 1), stride=(1, 1))
      (1): ReLU()
    )
  )
  (activation_post_process_1): HistogramObserver()
  (activation_post_process_2): HistogramObserver()
  (block2): Module(
    (linear): LinearReLU(
      (0): Linear(in_features=3, out_features=3, bias=True)
      (1): ReLU()
    )
    (conv2d): ConvReLU2d(
      (0): Conv2d(3, 3, kernel_size=(1, 1), stride=(1, 1))
      (1): ReLU()
    )
  )
  (activation_post_process_3): HistogramObserver()
  (activation_post_process_4): HistogramObserver()
  (activation_post_process_5): HistogramObserver()
)

```

In the above example, the module fqn of the first linear before the model was prepared was `block1.linear` but after prep (in the shown print out above) the module fqn of the same linear is `block1.linear.0`. Keep in mind that this module_fqn is found **not using the graph nodes, but `module.submodules()`** since finding node.target only gives a str so it's hard to tell if we actually have an instance of a supported class, which we can do with `module.submodules()`. 

However, we then need to get the actual node from the graph so we know what node to add observer after, and, for the exact same model as above, these are all the present nodes in the graph:
```
x
activation_post_process_0
block1.linear
activation_post_process_1
block1.conv2d
activation_post_process_2
block2.linear
activation_post_process_3
block2.conv2d
activation_post_process_4
<built-in function add>
<function relu at 0x7ff6e405e040>
activation_post_process_5
output
```

As you can see, there is no `block1.linear.0` there because it seems like the graph only contains the higher level node for fused modules (aka has `block1.linear` but not `block1.linear.0`) so that's when I check 1 level higher as well until we find a suitable parent one to put an observer in front of. Plus, since the linear / conv always come first in these fused modules when part of 1 [link to source](https://www.internalfb.com/code/fbsource/fbcode/caffe2/torch/nn/intrinsic/modules/fused.py?lines=35) Putting an observer in front of the fused module 1 level up is the same as putting it in front of them directly for the graph.

---

So, in the example you gave, the location of the linear is actually `submodule0.linear.0` and the relu would be `submodule0.linear.1`, and we wouldn't be able to find `submodule0.linear.0` in the graph, so we would go one level up and find `submodule0.linear` and put the ModelReportObserver before that, so it would be between the random_op and the fused `LinearRelu()`

HDCharles replied:
got it, you should explain this in comments thought, something like:

if the module we are looking for got fused, we will have module fqn for the linear of the fused module as
x.linear.0 but the graph will only have a node for the fused module which would have fqn x.linear so they will not match. To handle this, if we don't match, we then take off the last bit of the fqn i.e. foo.bar.baz -> foo.bar or in this case x.linear.0 -> x.linear and search again which will allow us to locate the correct module even in cases with fusion

vspenubarthi committed:
Update on "[ao] Implemented InputWeightEqualization Detector observer insertion"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This implements the functionality of adding observer points for the
input-weight equalization detector and contains the relavent tests for
this functionality. The full Detector functionality will be fleshed out
in a later commit.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles reviewed (approved):
looks good but you need to add a test case for the fusion case

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -529,8 +555,37 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+                    "insert_observer": obs_ctr,
+                    "insert_post": False,
+                    "observer_args": targeted_node.args,
+                }
```
This is starting to look a lot like [L277-L331](https://github.com/pytorch/pytorch/blob/21d51e9740c43038792cfd3487342bf3d7ded964/torch/ao/quantization/fx/_model_report/detector.py#L277-L331). I wonder if we should factor it in a common helper method that takes in the following args to reduce duplicate code:
```
supported_modules: List[...],
pre_observer_name: str,
post_observer_name: Optional[str]
```

vspenubarthi replied:
i think there is more clarity in the next commit in the stack, https://github.com/pytorch/pytorch/pull/80191, but this does diverge since the ModelReport Observer is instantiated in different ways within this loop depending on what the observer is, so I thought there was enough difference in implementation that it warranted being separate. That being said, I will definitely keep this in mind and focus more on refactoring where appropriate as I continue working on this.

vspenubarthi committed:
Update on "[ao] Implemented InputWeightEqualization Detector observer insertion"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This implements the functionality of adding observer points for the
input-weight equalization detector and contains the relavent tests for
this functionality. The full Detector functionality will be fleshed out
in a later commit.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
